File: compiler/interpreter.py
from ast_nodes import *

class SemanticError(Exception): pass
class RuntimeError(Exception): pass

class Interpreter:
    def __init__(self):
        self.env = {}
        self.output = []

    def eval(self, node):
        if isinstance(node, Num): return node.value
        if isinstance(node, Var):
            if node.name not in self.env: raise RuntimeError(f"Undefined variable '{node.name}'")
            return self.env[node.name]
        if isinstance(node, BinOp):
            a = self.eval(node.left)
            b = self.eval(node.right)
            t = node.op
            if t == 'PLUS': return a + b
            if t == 'MINUS': return a - b
            if t == 'STAR': return a * b
            if t == 'SLASH': return a // b
            if t == 'GT': return 1 if a > b else 0
            if t == 'LT': return 1 if a < b else 0
            if t == 'GE': return 1 if a >= b else 0
            if t == 'LE': return 1 if a <= b else 0
            if t == 'EQ': return 1 if a == b else 0
            raise RuntimeError('Unknown operator')
        raise RuntimeError('Unknown expression')

    def exec(self, node):
        if isinstance(node, VarDecl):
            self.env[node.name] = self.eval(node.expr)
        elif isinstance(node, AssignStmt):
            if node.name not in self.env:
                raise RuntimeError(f"Assign to undeclared variable '{node.name}'")
            self.env[node.name] = self.eval(node.expr)
        elif isinstance(node, PrintStmt):
            self.output.append(str(self.eval(node.expr)))
        elif isinstance(node, IfStmt):
            if self.eval(node.cond):
                for s in node.then_branch: self.exec(s)
            elif node.else_branch:
                for s in node.else_branch: self.exec(s)
        elif isinstance(node, ForStmt):
            self.exec(node.init)
            while self.eval(node.cond):
                for s in node.body: self.exec(s)
                name, expr = node.update
                # for-loop update always assigns, declare if missing
                self.env[name] = self.eval(expr)
        else:
            raise RuntimeError('Unknown statement')

    def interpret(self, program: Program):
        for stmt in program.statements:
            self.exec(stmt)

File: compiler/interpreter.ts
import {
  Program,
  Statement,
  VarDecl,
  AssignStmt,
  PrintStmt,
  IfStmt,
  ForStmt,
  Expression,
  BinOp,
  Num,
  Var,
} from './ast_nodes';

export class Interpreter {
  private env: Map<string, number> = new Map();
  output: string[] = [];

  interpret(program: Program): string {
    try {
      for (const stmt of program.statements) {
        this.execStatement(stmt);
      }
      return this.output.join('\n');
    } catch (e: any) {
      throw new Error(`Runtime error: ${e.message}`);
    }
  }

  private execStatement(stmt: Statement): void {
    if (stmt instanceof VarDecl) {
      const value = this.evalExpression(stmt.value);
      this.env.set(stmt.name, value);
    } else if (stmt instanceof AssignStmt) {
      const value = this.evalExpression(stmt.value);
      this.env.set(stmt.name, value);
    } else if (stmt instanceof PrintStmt) {
      const value = this.evalExpression(stmt.value);
      this.output.push(String(value));
    } else if (stmt instanceof IfStmt) {
      const condition = this.evalExpression(stmt.condition);
      if (condition) {
        for (const s of stmt.thenBlock) {
          this.execStatement(s);
        }
      } else if (stmt.elseBlock) {
        for (const s of stmt.elseBlock) {
          this.execStatement(s);
        }
      }
    } else if (stmt instanceof ForStmt) {
      this.execStatement(stmt.init);
      while (this.evalExpression(stmt.condition)) {
        for (const s of stmt.body) {
          this.execStatement(s);
        }
        this.execStatement(stmt.update);
      }
    }
  }

  private evalExpression(expr: Expression): number {
    if (expr instanceof Num) {
      return expr.value;
    } else if (expr instanceof Var) {
      const val = this.env.get(expr.name);
      if (val === undefined) {
        throw new Error(`Undefined variable: ${expr.name}`);
      }
      return val;
    } else if (expr instanceof BinOp) {
      const left = this.evalExpression(expr.left);
      const right = this.evalExpression(expr.right);
      switch (expr.op) {
        case '+':
          return left + right;
        case '-':
          return left - right;
        case '*':
          return left * right;
        case '/':
          if (right === 0) throw new Error('Division by zero');
          return Math.floor(left / right);
        case '>':
          return left > right ? 1 : 0;
        case '<':
          return left < right ? 1 : 0;
        case '>=':
          return left >= right ? 1 : 0;
        case '<=':
          return left <= right ? 1 : 0;
        case '==':
          return left === right ? 1 : 0;
        default:
          throw new Error(`Unknown operator: ${expr.op}`);
      }
    }
    throw new Error('Unknown expression type');
  }
}
